cpp/include/raft/spatial/knn/detail/ball_cover/registers-inl.cuh

```diff
@@ -551,69 +601,275 @@ RAFT_KERNEL block_rbc_kernel_eps_csr_pass(const value_t* X_index,
+  const value_t eps2 = eps * eps;
+
+#pragma nounroll
+  for (uint32_t cur_k0 = 0; cur_k0 < n_landmarks; cur_k0 += WarpSize) {
+    // Pre-compute landmark_dist & triangularization checks for 32 iterations
+    const uint32_t lane_k        = cur_k0 + lid;
+    const value_t lane_R_dist_sq = lane_k < n_landmarks ? dfunc(local_x_ptr, R + lane_k * dim, dim)
+                                                        : std::numeric_limits<value_idx>::max();
+    const int lane_check         = lane_k < n_landmarks
+                                     ? static_cast<int>(lane_R_dist_sq <= squared(eps + R_radius[lane_k]))
+                                     : 0;
+
+    int lane_mask = raft::ballot(lane_check);
+    if (lane_mask == 0) continue;
+
+    // reverse to use __clz instead of __ffs
+    lane_mask = __brev(lane_mask);
+    do {
+      // look for next k_offset
+      const uint32_t k_offset = __clz(lane_mask);
+
+      const uint32_t cur_k = cur_k0 + k_offset;
+
+      // The whole warp should iterate through the elements in the current R
+      const value_idx R_start_offset = R_indptr[cur_k];
+
+      // update lane_mask for next iteration - erase bits up to k_offset
+      lane_mask &= (0x7fffffff >> k_offset);
+
+      const uint32_t R_size = R_indptr[cur_k + 1] - R_start_offset;
+
+      // we have precomputed the query<->landmark distance
+      const value_t cur_R_dist = raft::sqrt(raft::shfl(lane_R_dist_sq, k_offset));
+
+      const uint32_t limit = Pow2<WarpSize>::roundDown(R_size);
+      uint32_t i           = limit + lid;
+
+      // R_1nn_dists are sorted ascendingly for each landmark
+      // Iterating backwards, after pruning the first point w.r.t. triangle
+      // inequality all subsequent points can be pruned as well
+      const value_t* y_ptr = X_reordered + (dim * (R_start_offset + i));
+      {
+        const value_t min_warp_dist =
+          limit < R_size ? R_1nn_dists[R_start_offset + limit] : cur_R_dist;
+        const value_t dist =
+          (i < R_size) ? dfunc(local_x_ptr, y_ptr, dim) : std::numeric_limits<value_idx>::max();
+        const bool in_range = (dist <= eps2);
+        if constexpr (write_pass) {
+          const int mask = raft::ballot(in_range);
+          if (in_range) {
+            const uint32_t index   = R_1nn_cols[R_start_offset + i];
+            const uint32_t row_pos = __popc(mask & lid_mask);
+            adj_ja[row_pos]        = index;
+          }
+          adj_ja += __popc(mask);
+        } else {
+          column_index_offset += (in_range);
+        }
+        // abort in case subsequent points cannot possibly be in reach
+        i *= (cur_R_dist - min_warp_dist <= eps);
+      }
 
-      const value_t* y_ptr = X_index + (n_cols * cur_candidate_ind);
-      if (dfunc(x_ptr, y_ptr, n_cols) <= eps) {
-        auto row_pos = atomicAdd(&column_index_smem, 1);
-        if (pass2) adj_ja[row_pos] = cur_candidate_ind;
+      while (i >= WarpSize) {
```

Comment:
Oh, one more thing. I was trying to make a reproducer for the "branch on divergence" (`BRA.DIV`) instructions that we've been seeing, and it occurred to me that the loop condition of the inner-most loop depends on the lane id, because `i` is `limit + lid`, but iiuc the possible out-of-bounds are handled before the loop, and all lanes always execute that while loop together.

So one thing to try would be to help the compiler know that all lanes are active: for example broadcast `i0`, use it in the loop condition, and inside the loop use `i = i0 + lid`.

Reply:
Erm, unfortunately that idea does not work well with the early exit:

```c++
// abort in case subsequent points cannot possibly be in reach
i0 *= (cur_R_dist - min_warp_dist <= eps);
```

Reply:
You can probably implement that with a `VOTE`.

Replace:
```c++
// abort in case subsequent points cannot possibly be in reach
i *= (cur_R_dist - min_warp_dist <= eps);
```

with:
```
next_loop = (i >= WarpSize) && (cur_R_dist - min_warp_dist <= eps);
```

and change loop condition to:

```c++
while (raft::any(next_loop))
```